# Import Libraries

In [1]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

## Data Transformations

We first start with defining our data transformations. We need to think what our data is and how can we augment it to correct represent images which it might not see otherwise.


In [2]:
# Train Phase transformations
train_transforms = transforms.Compose([
                                       transforms.RandomRotation((-7.0, 7.0), fill=(1,)),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])

# Test Phase transformations
test_transforms = transforms.Compose([
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])


# Dataset and Creating Train/Test Split

In [3]:
train = datasets.MNIST('./data', train=True, download=True, transform=train_transforms)
test = datasets.MNIST('./data', train=False, download=True, transform=test_transforms)

100%|██████████| 9912422/9912422 [00:00<00:00, 147820020.85it/s]

Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw


100%|██████████| 28881/28881 [00:00<00:00, 111748795.04it/s]


Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw



100%|██████████| 1648877/1648877 [00:00<00:00, 37791343.25it/s]

Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw


100%|██████████| 4542/4542 [00:00<00:00, 21772032.88it/s]


Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw



# Dataloader Arguments & Test/Train Dataloaders


In [4]:
SEED = 1

# CUDA?
cuda = torch.cuda.is_available()
print("CUDA Available?", cuda)

# For reproducibility
torch.manual_seed(SEED)

if cuda:
    torch.cuda.manual_seed(SEED)

# dataloader arguments - something you'll fetch these from cmdprmt
dataloader_args = dict(shuffle=True, batch_size=128, num_workers=4, pin_memory=True) if cuda else dict(shuffle=True, batch_size=64)

# train dataloader
train_loader = torch.utils.data.DataLoader(train, **dataloader_args)

# test dataloader
test_loader = torch.utils.data.DataLoader(test, **dataloader_args)

CUDA Available? True


/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:560: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


# The model
Let's start with the model we first saw

In [5]:
from models import model_v2

# Model Params
Can't emphasize on how important viewing Model Summary is.
Unfortunately, there is no in-built model visualizer, so we have to take external help

In [6]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print(device)
model = model_v2().to(device)
summary(model, input_size=(1, 28, 28))

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 26, 26]             144
              ReLU-2           [-1, 16, 26, 26]               0
       BatchNorm2d-3           [-1, 16, 26, 26]              32
           Dropout-4           [-1, 16, 26, 26]               0
            Conv2d-5           [-1, 20, 24, 24]           2,880
              ReLU-6           [-1, 20, 24, 24]               0
       BatchNorm2d-7           [-1, 20, 24, 24]              40
           Dropout-8           [-1, 20, 24, 24]               0
            Conv2d-9           [-1, 10, 24, 24]             200
        MaxPool2d-10           [-1, 10, 12, 12]               0
           Conv2d-11           [-1, 10, 12, 12]             900
             ReLU-12           [-1, 10, 12, 12]               0

# Training and Testing

All right, so we have 24M params, and that's too many, we know that. But the purpose of this notebook is to set things right for our future experiments.

Looking at logs can be boring, so we'll introduce **tqdm** progressbar to get cooler logs.

Let's write train and test functions

In [7]:
from utils import train, test

In [8]:
from torch.optim.lr_scheduler import StepLR

model =  model_v2().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
# scheduler = StepLR(optimizer, step_size=6, gamma=0.1)


EPOCHS = 21
for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    # scheduler.step()
    test(model, device, test_loader)

EPOCH: 0


Loss=0.09005427360534668 Batch_id=468 Accuracy=90.42: 100%|██████████| 469/469 [00:21<00:00, 21.52it/s]



Test set: Average loss: 0.0691, Accuracy: 9805/10000 (98.05%)

EPOCH: 1


Loss=0.05079081282019615 Batch_id=468 Accuracy=97.96: 100%|██████████| 469/469 [00:22<00:00, 21.07it/s]



Test set: Average loss: 0.0425, Accuracy: 9888/10000 (98.88%)

EPOCH: 2


Loss=0.07478278875350952 Batch_id=468 Accuracy=98.41: 100%|██████████| 469/469 [00:21<00:00, 21.86it/s]



Test set: Average loss: 0.0414, Accuracy: 9882/10000 (98.82%)

EPOCH: 3


Loss=0.08886492997407913 Batch_id=468 Accuracy=98.70: 100%|██████████| 469/469 [00:20<00:00, 22.78it/s]



Test set: Average loss: 0.0355, Accuracy: 9898/10000 (98.98%)

EPOCH: 4


Loss=0.05365810915827751 Batch_id=468 Accuracy=98.74: 100%|██████████| 469/469 [00:20<00:00, 22.51it/s]



Test set: Average loss: 0.0393, Accuracy: 9887/10000 (98.87%)

EPOCH: 5


Loss=0.04258883371949196 Batch_id=468 Accuracy=98.88: 100%|██████████| 469/469 [00:21<00:00, 21.54it/s]



Test set: Average loss: 0.0328, Accuracy: 9897/10000 (98.97%)

EPOCH: 6


Loss=0.025998430326581 Batch_id=468 Accuracy=98.90: 100%|██████████| 469/469 [00:23<00:00, 20.30it/s]



Test set: Average loss: 0.0291, Accuracy: 9910/10000 (99.10%)

EPOCH: 7


Loss=0.09406533092260361 Batch_id=468 Accuracy=99.05: 100%|██████████| 469/469 [00:22<00:00, 21.18it/s]



Test set: Average loss: 0.0297, Accuracy: 9912/10000 (99.12%)

EPOCH: 8


Loss=0.08142108470201492 Batch_id=468 Accuracy=99.07: 100%|██████████| 469/469 [00:22<00:00, 21.10it/s]



Test set: Average loss: 0.0250, Accuracy: 9919/10000 (99.19%)

EPOCH: 9


Loss=0.06215141713619232 Batch_id=468 Accuracy=99.09: 100%|██████████| 469/469 [00:21<00:00, 21.79it/s]



Test set: Average loss: 0.0247, Accuracy: 9921/10000 (99.21%)

EPOCH: 10


Loss=0.07050184905529022 Batch_id=468 Accuracy=99.09: 100%|██████████| 469/469 [00:20<00:00, 22.56it/s]



Test set: Average loss: 0.0239, Accuracy: 9927/10000 (99.27%)

EPOCH: 11


Loss=0.044766321778297424 Batch_id=468 Accuracy=99.12: 100%|██████████| 469/469 [00:21<00:00, 22.08it/s]



Test set: Average loss: 0.0262, Accuracy: 9913/10000 (99.13%)

EPOCH: 12


Loss=0.01309166382998228 Batch_id=468 Accuracy=99.20: 100%|██████████| 469/469 [00:21<00:00, 21.41it/s]



Test set: Average loss: 0.0244, Accuracy: 9928/10000 (99.28%)

EPOCH: 13


Loss=0.12477622181177139 Batch_id=468 Accuracy=99.22: 100%|██████████| 469/469 [00:21<00:00, 21.42it/s]



Test set: Average loss: 0.0240, Accuracy: 9921/10000 (99.21%)

EPOCH: 14


Loss=0.00964284222573042 Batch_id=468 Accuracy=99.28: 100%|██████████| 469/469 [00:21<00:00, 21.51it/s]



Test set: Average loss: 0.0246, Accuracy: 9921/10000 (99.21%)

EPOCH: 15


Loss=0.06722014397382736 Batch_id=468 Accuracy=99.31: 100%|██████████| 469/469 [00:20<00:00, 22.36it/s]



Test set: Average loss: 0.0213, Accuracy: 9931/10000 (99.31%)

EPOCH: 16


Loss=0.010719981044530869 Batch_id=468 Accuracy=99.27: 100%|██████████| 469/469 [00:20<00:00, 22.51it/s]



Test set: Average loss: 0.0203, Accuracy: 9933/10000 (99.33%)

EPOCH: 17


Loss=0.01570579968392849 Batch_id=468 Accuracy=99.30: 100%|██████████| 469/469 [00:21<00:00, 21.82it/s]



Test set: Average loss: 0.0215, Accuracy: 9930/10000 (99.30%)

EPOCH: 18


Loss=0.06970813125371933 Batch_id=468 Accuracy=99.24: 100%|██████████| 469/469 [00:21<00:00, 21.64it/s]



Test set: Average loss: 0.0227, Accuracy: 9929/10000 (99.29%)

EPOCH: 19


Loss=0.003820765996351838 Batch_id=468 Accuracy=99.34: 100%|██████████| 469/469 [00:21<00:00, 21.50it/s]



Test set: Average loss: 0.0202, Accuracy: 9936/10000 (99.36%)

EPOCH: 20


Loss=0.012009280733764172 Batch_id=468 Accuracy=99.31: 100%|██████████| 469/469 [00:22<00:00, 21.26it/s]



Test set: Average loss: 0.0198, Accuracy: 9929/10000 (99.29%)



## Target :- 
**1. Add Augmentation** 

**2. Results :**

        1. Parameters : 8K
        2. Best Train Accuracy : 99.34 %
        3. Best Test Accuracy : 99.36 % (19th epoch)

**3. Analysis :**

        1. Model has ran for 20 epochs
        2. Due to augmentation and dropout layers model is training hard, so there is scope of learning more